src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Windows.cs

```diff
@@ -2,35 +2,196 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
 using System.Threading;
 
 namespace Microsoft.Win32.SafeHandles
 {
-    public sealed class SafeFileHandle : SafeHandleZeroOrMinusOneIsInvalid
+    public sealed partial class SafeFileHandle : SafeHandleZeroOrMinusOneIsInvalid
     {
-        private bool? _isAsync;
+        internal const FileOptions NoBuffering = (FileOptions)0x20000000;
+        private FileOptions _fileOptions = (FileOptions)(-1);
+        private int _fileType = -1;
 
         public SafeFileHandle() : base(true)
         {
-            _isAsync = null;
         }
 
         public SafeFileHandle(IntPtr preexistingHandle, bool ownsHandle) : base(ownsHandle)
         {
             SetHandle(preexistingHandle);
-
-            _isAsync = null;
         }
 
-        internal bool? IsAsync
+        private SafeFileHandle(IntPtr preexistingHandle, bool ownsHandle, FileOptions fileOptions) : base(ownsHandle)
         {
-            get => _isAsync;
-            set => _isAsync = value;
+            SetHandle(preexistingHandle);
+
+            _fileOptions = fileOptions;
         }
 
+        public bool IsAsync => (GetFileOptions() & FileOptions.Asynchronous) != 0;
+
+        internal bool CanSeek => !IsClosed && GetFileType() == Interop.Kernel32.FileTypes.FILE_TYPE_DISK;
+
+        internal bool IsPipe => GetFileType() == Interop.Kernel32.FileTypes.FILE_TYPE_PIPE;
+
         internal ThreadPoolBoundHandle? ThreadPoolBinding { get; set; }
 
-        protected override bool ReleaseHandle() =>
-            Interop.Kernel32.CloseHandle(handle);
+        internal static unsafe SafeFileHandle Open(string fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options, long preallocationSize)
+        {
+            using (DisableMediaInsertionPrompt.Create())
+            {
+                SafeFileHandle fileHandle = new SafeFileHandle(
+                    NtCreateFile(fullPath, mode, access, share, options, preallocationSize),
+                    ownsHandle: true,
+                    options);
+
+                fileHandle.InitThreadPoolBindingIfNeeded();
+
+                return fileHandle;
+            }
+        }
+
+        private static IntPtr NtCreateFile(string fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options, long preallocationSize)
+        {
+            uint ntStatus;
+            IntPtr fileHandle;
+
+            const string MandatoryNtPrefix = @"\??\";
+            if (fullPath.StartsWith(MandatoryNtPrefix, StringComparison.Ordinal))
+            {
+                (ntStatus, fileHandle) = Interop.NtDll.NtCreateFile(fullPath, mode, access, share, options, preallocationSize);
+            }
+            else
+            {
+                var vsb = new ValueStringBuilder(stackalloc char[1024]);
```

Comment:
FWIW, 2K is generally more than we like to stackalloc.

Reply:
Would 256 chars be ok?

Reply:
Yes, that's commonly what we use (either explicitly or somewhat implicitly from using MAX_PATH).